devnet-sdk/testing/systest/multi_client.go

```diff
@@ -15,47 +15,77 @@ import (
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/ethclient"
 	"github.com/ethereum/go-ethereum/log"
-	"github.com/stretchr/testify/require"
 )
 
+type HeaderProvider interface {
+	BlockByNumber(ctx context.Context, number *big.Int) (*types.Block, error)
+	HeaderByNumber(ctx context.Context, number *big.Int) (*types.Header, error)
+	HeaderByHash(ctx context.Context, hash common.Hash) (*types.Header, error)
+	Close()
+}
+
+var _ HeaderProvider = (*ethclient.Client)(nil)
+
+func getEthClients(chain system.Chain) ([]HeaderProvider, error) {
+	hps := make([]HeaderProvider, 0, len(chain.Nodes()))
+	for _, n := range chain.Nodes() {
+		gethCl, err := n.GethClient()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get geth client: %w", err)
+		}
+		hps = append(hps, gethCl)
+	}
+	return hps, nil
+}
+
 // RequireNoChainFork checks that the L2 chain has not forked now, and returns a
 // function that check again (to be called at the end of the test).
-func RequireNoChainFork(t T, chain system.L2Chain, logger log.Logger) func() {
-	ctx := t.Context()
-
-	clients := make([]*ethclient.Client, 0, len(chain.Nodes()))
-	for _, node := range chain.Nodes() {
-		client, err := node.GethClient()
-		require.NoError(t, err)
-		clients = append(clients, client)
+func RequireNoChainFork(ctx context.Context, chain system.L2Chain, logger log.Logger) (func() error, error) {
+	clients, err := getEthClients(chain)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get eth clients: %w", err)
 	}
+	return requireNoChainFork(ctx, clients, logger)
+}
+
+// requireNoChainFork checks that the L2 chain has not forked now, and returns a
+// function that check again (to be called at the end of the test).
+func requireNoChainFork(ctx context.Context, clients []HeaderProvider, logger log.Logger) (func() error, error) {
 
-	// We use a multiclient where feasible to automatically check for consistency
+	// We use a multiclient to automatically check for consistency
 	// between the nodes.
 	l2MultiClient := NewMultiClient(clients)
 
 	// Setup chain fork detection
-	logger.Info("Setting up chain fork detection")
+	logger.Info("Running first chain fork detection check")
 	l2StartHeader, err := l2MultiClient.HeaderByNumber(ctx, nil)
-	require.NoError(t, err)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get L2 start block: %w", err)
+	}
 	logger.Debug("Got L2 head block", "number", l2StartHeader.Number)
-	return func() {
+	return func() error {
+		logger.Info("Running second chain fork detection check")
 		l2EndHeader, err := l2MultiClient.HeaderByNumber(ctx, nil)
-		require.NoError(t, err)
-		require.True(t, l2EndHeader.Number.Cmp(l2StartHeader.Number) > 0, "L2 chain should have progressed")
+		if err != nil {
+			return fmt.Errorf("failed to get L2 end block: %w", err)
+		}
 		logger.Debug("Got L2 end block", "number", l2EndHeader.Number)
-	}
+		if !(l2EndHeader.Number.Int64() > l2StartHeader.Number.Int64()) {
```

Comment:
Side note: This seems safe, even producing 5 blocks per second for a million years you wouldn't get close to running out of block numbers. So why are big.Ints used by go-ethereum to begin with?